File: vercel-api/tests/fefo-allocation.test.ts
import {describe, expect, it} from "vitest";
import {
  allocateFefo,
  InsufficientStockError,
  sortBatchesForFefo,
  type FefoBatchInput,
} from "../src/services/inventory/fefo-allocation";

const now = new Date("2026-07-22T12:00:00.000Z");
const timezone = "Africa/Freetown";

function batch(
  partial: Partial<FefoBatchInput> & Pick<FefoBatchInput, "id">,
): FefoBatchInput {
  return {
    quantityRemaining: 10,
    unitCostMinor: 100,
    expiryDateKnown: false,
    expiryDate: null,
    status: "active",
    ...partial,
  };
}

describe("allocateFefo", () => {
  it("allocates earliest known expiry first", () => {
    const allocations = allocateFefo({
      requestedQty: 3,
      timezone,
      now,
      batches: [
        batch({
          id: "late",
          quantityRemaining: 10,
          unitCostMinor: 200,
          expiryDateKnown: true,
          expiryDate: "2026-12-01",
        }),
        batch({
          id: "early",
          quantityRemaining: 10,
          unitCostMinor: 150,
          expiryDateKnown: true,
          expiryDate: "2026-08-01",
        }),
      ],
    });

    expect(allocations).toEqual([
      {
        batchId: "early",
        quantity: 3,
        unitCostMinor: 150,
        expiryDate: "2026-08-01",
        lineCostMinor: 450,
      },
    ]);
  });

  it("spans two batches when the first is insufficient", () => {
    const allocations = allocateFefo({
      requestedQty: 7,
      timezone,
      now,
      batches: [
        batch({
          id: "a",
          quantityRemaining: 4,
          unitCostMinor: 100,
          expiryDateKnown: true,
          expiryDate: "2026-08-01",
        }),
        batch({
          id: "b",
          quantityRemaining: 10,
          unitCostMinor: 120,
          expiryDateKnown: true,
          expiryDate: "2026-09-01",
        }),
      ],
    });

    expect(allocations).toEqual([
      {
        batchId: "a",
        quantity: 4,
        unitCostMinor: 100,
        expiryDate: "2026-08-01",
        lineCostMinor: 400,
      },
      {
        batchId: "b",
        quantity: 3,
        unitCostMinor: 120,
        expiryDate: "2026-09-01",
        lineCostMinor: 360,
      },
    ]);
  });

  it("places unknown-expiry batches last", () => {
    const ordered = sortBatchesForFefo([
      batch({
        id: "unknown",
        expiryDateKnown: false,
        expiryDate: null,
      }),
      batch({
        id: "known",
        expiryDateKnown: true,
        expiryDate: "2026-10-01",
      }),
    ]);
    expect(ordered.map((row) => row.id)).toEqual(["known", "unknown"]);

    const allocations = allocateFefo({
      requestedQty: 5,
      timezone,
      now,
      batches: [
        batch({
          id: "unknown",
          quantityRemaining: 20,
          unitCostMinor: 50,
          expiryDateKnown: false,
          expiryDate: null,
        }),
        batch({
          id: "known",
          quantityRemaining: 2,
          unitCostMinor: 80,
          expiryDateKnown: true,
          expiryDate: "2026-08-15",
        }),
      ],
    });

    expect(allocations.map((row) => row.batchId)).toEqual([
      "known",
      "unknown",
    ]);
    expect(allocations[0]?.quantity).toBe(2);
    expect(allocations[1]?.quantity).toBe(3);
    expect(allocations[1]?.expiryDate).toBeNull();
  });

  it("throws when saleable stock is insufficient", () => {
    expect(() =>
      allocateFefo({
        requestedQty: 5,
        timezone,
        now,
        batches: [
          batch({
            id: "small",
            quantityRemaining: 2,
            expiryDateKnown: true,
            expiryDate: "2026-08-01",
          }),
          batch({
            id: "expired-status",
            quantityRemaining: 50,
            status: "expired",
            expiryDateKnown: true,
            expiryDate: "2026-01-01",
          }),
          batch({
            id: "past-date",
            quantityRemaining: 50,
            status: "active",
            expiryDateKnown: true,
            expiryDate: "2026-07-01",
          }),
          batch({
            id: "depleted",
            quantityRemaining: 0,
            status: "depleted",
          }),
          batch({
            id: "voided",
            quantityRemaining: 10,
            status: "voided",
          }),
        ],
      }),
    ).toThrow(InsufficientStockError);
  });
});

File: test/features/inventory/fefo_allocation_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:sabibom/features/inventory/domain/inventory_batch.dart';
import 'package:sabibom/features/inventory/domain/product_profit_calculator.dart';

List<InventoryBatch> orderForFefo(List<InventoryBatch> batches) {
  final eligible = batches
      .where((batch) => batch.status == InventoryBatchStatus.active)
      .where((batch) => batch.quantityRemaining > 0)
      .toList();
  eligible.sort((left, right) {
    final leftKnown = left.hasKnownExpiry;
    final rightKnown = right.hasKnownExpiry;
    if (leftKnown && !rightKnown) return -1;
    if (!leftKnown && rightKnown) return 1;
    if (leftKnown && rightKnown) {
      return left.expiryDate!.compareTo(right.expiryDate!);
    }
    return left.receivedAt.compareTo(right.receivedAt);
  });
  return eligible;
}

List<BatchAllocation> allocateFefo({
  required double quantity,
  required List<InventoryBatch> batches,
}) {
  var remaining = quantity;
  final allocations = <BatchAllocation>[];
  for (final batch in orderForFefo(batches)) {
    if (remaining <= 0) break;
    final take = remaining < batch.quantityRemaining
        ? remaining
        : batch.quantityRemaining;
    allocations.add(
      BatchAllocation(
        batchId: batch.id,
        quantity: take,
        unitCostMinor: batch.unitCostMinor,
        expiryDate: batch.expiryDate,
        lineCostMinor: ProductProfitCalculator.lineValue(
          take,
          batch.unitCostMinor,
        ),
      ),
    );
    remaining -= take;
  }
  if (remaining > 0.0000001) {
    throw StateError('Insufficient batch stock');
  }
  return allocations;
}

void main() {
  test('FEFO consumes earliest known expiry first and spans batches', () {
    final batches = <InventoryBatch>[
      InventoryBatch(
        id: 'b',
        businessId: 'biz',
        branchId: 'main',
        productId: 'milk',
        productName: 'Milk',
        sourceType: InventoryBatchSourceType.purchase,
        quantityReceived: 10,
        quantityRemaining: 10,
        unitCostMinor: 2200,
        sellingPriceAtReceiptMinor: 3000,
        expiryDateKnown: true,
        expiryDate: DateTime.utc(2026, 9, 20),
        receivedAt: DateTime.utc(2026, 7, 1),
        status: InventoryBatchStatus.active,
      ),
      InventoryBatch(
        id: 'a',
        businessId: 'biz',
        branchId: 'main',
        productId: 'milk',
        productName: 'Milk',
        sourceType: InventoryBatchSourceType.purchase,
        quantityReceived: 5,
        quantityRemaining: 5,
        unitCostMinor: 2000,
        sellingPriceAtReceiptMinor: 3000,
        expiryDateKnown: true,
        expiryDate: DateTime.utc(2026, 8, 10),
        receivedAt: DateTime.utc(2026, 7, 1),
        status: InventoryBatchStatus.active,
      ),
      InventoryBatch(
        id: 'unknown',
        businessId: 'biz',
        branchId: 'main',
        productId: 'milk',
        productName: 'Milk',
        sourceType: InventoryBatchSourceType.manualStockIn,
        quantityReceived: 4,
        quantityRemaining: 4,
        unitCostMinor: 2100,
        sellingPriceAtReceiptMinor: 3000,
        expiryDateKnown: false,
        receivedAt: DateTime.utc(2026, 7, 2),
        status: InventoryBatchStatus.active,
      ),
    ];

    final allocations = allocateFefo(quantity: 7, batches: batches);
    expect(allocations.map((a) => a.batchId).toList(), <String>['a', 'b']);
    expect(allocations[0].quantity, 5);
    expect(allocations[1].quantity, 2);
    expect(orderForFefo(batches).last.id, 'unknown');
  });

  test('FEFO rejects insufficient stock', () {
    expect(
      () => allocateFefo(
        quantity: 3,
        batches: <InventoryBatch>[
          InventoryBatch(
            id: 'a',
            businessId: 'biz',
            branchId: 'main',
            productId: 'milk',
            productName: 'Milk',
            sourceType: InventoryBatchSourceType.purchase,
            quantityReceived: 2,
            quantityRemaining: 2,
            unitCostMinor: 2000,
            sellingPriceAtReceiptMinor: 3000,
            expiryDateKnown: true,
            expiryDate: DateTime.utc(2026, 8, 10),
            receivedAt: DateTime.utc(2026, 7, 1),
            status: InventoryBatchStatus.active,
          ),
        ],
      ),
      throwsStateError,
    );
  });
}
